some upgrades to game validation, rich text, property inheritance, and extension methods

brine opened the pull request:
- proper handling and inheritance of null property values for string and richtext custom properties
- card properties in the set XML can use the value attribute OR element inner text, it is no longer restricted based on property type (string vs richText)
- a new python API method, card.hasProperty(), to detect if a property value is null or missing
- added summary comments in card and game extension methods
- restored some deprecated methods and added warnings.  This was to prevent some changes from breaking existing deck editor plugins and other apps using our libraries

brine commented on octgnFX/Octgn/Scripting/Versions/Script_3_1_0_2.cs:
```diff
@@ -525,7 +525,7 @@ public object CardProperty(int id, string property)
             Card c = Card.Find(id);
             property = property.ToLowerInvariant();
-            var value = c.GetProperty(property, "", StringComparison.InvariantCultureIgnoreCase, c.Alternate());
+            var value = c.GetProperty(property, null, StringComparison.InvariantCultureIgnoreCase, c.Alternate());
```
this is a major point of concern for me.  I'm really not a fan of this default value being hard-set in the API.  Changing it to null has huge changes for existing python scripts (as card.property == "" was the standard way of checking if a card property is missing).

Maybe we should keep this as "" for now and implement it properly in the next API revision?

kellyelton replied:
I agree, alternatively, we could spin up 3_1_0_3 and make the change in there.

kellyelton replied:
Alternatively we could add a property to the gamedef somewhere that's like SetDefaultCardPropertyToNull="true" to enable it if someone wants it.

brine replied:
I think if we go down that route it'd be better to just make a card.getProperty() call that does the proper method and then implement it properly in 3.1.0.3

kellyelton replied:
agreed

brine commented:
OK I've reverted python's default value for null properties back to "", and added a new API call `card.hasProperty()` to check if the property is defined on the card.  This should keep compatibility while still providing the new null property setup for games that want to add it.

brine commented:
sorry kelly this kinda turned into a mega pull request again.
